astroid/inference.py

```diff
@@ -1278,3 +1307,128 @@ def infer_functiondef(
 
 
 nodes.FunctionDef._infer = infer_functiondef
+
+
+# pylint: disable-next=too-many-return-statements
+def _infer_node(
+    node: nodes.NodeNG, context: InferenceContext | None = None, **kwargs: Any
+) -> Generator[InferenceResult, None, None]:
+    """Find the infer method for the given node and call it."""
+    if isinstance(
```

Comment:
Given the number of time this is going to be called it might be a good idea to do stats on the most common one so it's tested first if we go with this approach.

Reply:
Yeah I think we might want to refactor this anyway before releasing the final version.

For now I'd say, let's use this as a start and see how we can optimise. I think there are a lot of other small optimisations we could do here.

Reply:
Even with a statistical approach, I'm nervous about this approach, it feels like a code smell that we should be using polymorphism, which is more like the existing way.

Reply:
We could also use a dictionary of nodes types and associated methods, which makes it (almost) `O(1)`. However, I think there might be good reason to eventually refactor this inference function to be part of an instantiated `AstroidManager` object instead of having it rely on one global `AstroidManager` class (as this makes parallelisation much harder).
Thus, too many optimisations to this don't make sense as the moment I think. It's more about the intent of decoupling nodes from the interaction with nodes.

Reply:
Agree with @jacobtylerwalls about the code smell. Couldn't we put this in the nodes instead of outside the node ? Sometime the deferred creation of the infer function wasn't required.

Reply:
Just look at the imports that are needed in this file. If we want to put this in `astroid.nodes` we would have to do all of those within the methods. See also my other comment in the main thread.
This design really boils down whether the nodes should know how to infer themselves. I don't think they do as it heavily complicates the structure of the code and so far we haven't found a good way to combine it in one file.